File: tools/syxtobin/syxtobin.cpp
#include "syxtobin.h"

void print_vector_error(int i, const char* error, std::vector<byte> v) {
	if (v.size() == 0) fprintf(stderr, "Runtime error at position %08x: can't expect empty vector %s.", i, error);
	else {
		fprintf(stderr, "Parse error at position %08x: expected %s (%02x", i, error, v[0]);

		for (int j = 1; j < v.size(); j++)
			fprintf(stderr, ", %02x", v[j]);

		fprintf(stderr, ").\n");
	}

	exit(5);
}

void verify_byte(int i, byte value, const char* error) {
	if (input.data[i] != value) {
		fprintf(stderr, "Parse error at position %08x: expected %s (%02x).\n", i, error, value);
		exit(5);
	}
}

void verify_byte(int i, const std::vector<byte> values, const char* error) {
	for (int j = 0; j < values.size(); j++)
		if (input.data[i] == values[j]) return;

	print_vector_error(i, error, values);
	exit(5);
}

int verify_byte_array(int i, const std::vector<byte> values, const char* error) {
	for (int j = 0; j < values.size(); j++)
		if (input.data[i + j] != values[j]) {
			fprintf(stderr, "Parse error at position %08x: expected %s (", i, error);

			exit(5);
		}

	return values.size();
}

uint nibbles_to_uint(int* i, int length) {
	uint result = 0;

	for (int j = 0; j < length; j++) {
		result <<= 0x04;
		result |= input.data[(*i)++] & 0x0F;
	}

	return result;
}

void verify_uints(int i, uint a, uint b, const char* error) {
	if (a != b) {
		fprintf(stderr, "Syntax error at position %08x: mismatching %s.\n", i, error);
		exit(5);
	}
}

byte convert_syxtobin() {
	std::vector<byte> expected_types = {UPDATE_INIT};

	byte family = LPX_FAMILY_ID;
	byte target = LPX_PRODUCT_ID;
	uint version = 0;
	uint checksum = 0;
	uint block = 0;

	for (int i = 0; i < input.size;) {
		verify_byte(i++, SYSEX_START, "SYSEX_START");

		i += verify_byte_array(i, novation_header, "NOVATION_HEADER");

		verify_byte(i, expected_types, "EXPECTED_TYPES");

		switch (input.data[i++]) {
			case UPDATE_INIT:
				verify_byte(i, families, "LP_FAMILY_ID");
				family = input.data[i++];

				verify_byte(i, *get_products(family), "LP_PRODUCT_ID");
				target = input.data[i++];

				version = nibbles_to_uint(&i, 6);

				if (family == LPX_FAMILY_ID) expected_types = {UPDATE_HEADER};
				else if (family == LPRGB_FAMILY_ID) {
					if (!allocate_buffer(&output, BLOCK_SIZE_BYTES, "OUTPUT")) exit(4);

					block = 1;
					expected_types = {UPDATE_WRITE, UPDATE_FINISH};
				}

				break;

			case UPDATE_HEADER:
				i += 1;

				verify_uints(i, version, nibbles_to_uint(&i, 6), "VERSION");

				if (!allocate_buffer(&output, nibbles_to_uint(&i, 8), "OUTPUT")) exit(4);

				checksum = nibbles_to_uint(&i, 8);

				block = 1;
				expected_types = {UPDATE_WRITE, UPDATE_FINISH};
				break;

			case UPDATE_FINISH:
				block = 0;

				if (family == LPX_FAMILY_ID) expected_types = {};
				else if (family == LPRGB_FAMILY_ID)	expected_types = {UPDATE_FOOTER};

			case UPDATE_WRITE:
				if (family == LPRGB_FAMILY_ID && block > 0) {
					if (!reallocate_buffer(&output, BLOCK_SIZE_BYTES, "firmware conversion")) exit(5);
				}

				for (int j = 0; j < BLOCK_SIZE_BITS; j++) {
					int shift = 6 - j % 7;
					int target = block * BLOCK_SIZE_BYTES + j / 8;

					if (family == LPX_FAMILY_ID && target >= output.size) {
						expected_types = {UPDATE_FINISH};
						break;
					}

					if (j % 8 == 0) output.data[target] = 0;
					
					output.data[target] |= (input.data[i + j / 7] & (1 << shift)) >> shift << (7 - j % 8);
				}

				block++;
				i += BLOCK_SIZE_7BITS;
				break;

			case UPDATE_FOOTER:
				i += 1;
				i += verify_byte_array(i, rgb_firmware_footer, "RGB_FIRMWARE_FOOTER");
				i += 8;

				expected_types = {};
				break;
		}

		verify_byte(i++, SYSEX_END, "SYSEX_END");
	}

	if (family == LPX_FAMILY_ID) verify_uints(input.size, checksum, crc32(&output), "CHECKSUM");

	return target;
}

File: tools_rust/src/bin/syxtobin.rs
use std::{path::PathBuf, option::Option};
use structopt::StructOpt;
use tools_rust::*;

pub struct SyxToBin {
    input: Vec<u8>,
    output: Vec<u8>,
    expected_types: Vec<u8>,
    family: u8,
    target: u8,
    block: usize
}

impl SyxToBin {
    fn new(input: Vec<u8>) -> SyxToBin {
        SyxToBin {
            input,
            output: vec![],
            expected_types: vec![UPDATE_INIT],
            family: LPX_FAMILY_ID,
            target: LPX_PRODUCT_ID,
            block: 0,
        }
    }
    
    fn convert(&mut self) -> &Vec<u8> {
        let mut i: usize = 0;
        let mut version: u32 = 0;
        let mut checksum: u32 = 0;
        
        self.output = Vec::new();
        
        while i < self.input.len() {
            i += self.verify_byte(i, SYSEX_START, "SYSEX_START");
            i += self.verify_byte_array(i, &NOVATION_HEADER.to_vec(), "NOVATION_HEADER");
            i += self.verify_byte_in_array(i, &self.expected_types, "EXPECTED_TYPES");
            
            match self.input[i - 1] {
                UPDATE_INIT => {
                    i += self.verify_byte_in_array(i, &FAMILIES.to_vec(), "LP_FAMILY_ID");
                    self.family = self.input[i - 1];
                    
                    i += self.verify_byte_in_array(i, &get_products(self.family), "LP_PRODUCT_ID");
                    self.target = self.input[i - 1];
                    
                    version = self.nibbles_to_u32(&mut i, 6);
                    
                    if self.family == LPX_FAMILY_ID {
                        self.expected_types = vec![UPDATE_HEADER];
                    }
                    else if self.family == LPRGB_FAMILY_ID {
                        self.output.append(&mut vec![0; BLOCK_SIZE_BYTES]);
                        
                        self.block = 1;
                        self.expected_types = vec![UPDATE_WRITE, UPDATE_FINISH];
                    }
                },
                UPDATE_HEADER => {
                    i += 1;
                    
                    verify_uints(i, version, self.nibbles_to_u32(&mut i, 6), "VERSION");
                    
                    self.output.append(&mut vec![0; self.nibbles_to_u32(&mut i, 8) as usize]);
                    checksum = self.nibbles_to_u32(&mut i, 8);
                    
                    self.block = 1;
                    self.expected_types = vec![UPDATE_WRITE, UPDATE_FINISH];
                },
                UPDATE_FINISH => {
                    self.block = 0;
                    
                    if self.family == LPX_FAMILY_ID { self.expected_types = vec![] }
                    else if self.family == LPRGB_FAMILY_ID { self.expected_types = vec![UPDATE_FOOTER] }
                    
                    self.update_write(&mut i);
                },
                UPDATE_WRITE => self.update_write(&mut i),
                UPDATE_FOOTER => {
                    i += 1;
                    self.verify_byte_array(i, &RGB_FIRMWARE_FOOTER.to_vec(), "RGB_FIRMWARE_FOOTER");
                    
                    self.expected_types = vec![];
                }
                _ => {}
            }
            
            i += self.verify_byte(i, SYSEX_END, "SYSEX_END");
        }
        if self.family == LPX_FAMILY_ID {
            verify_uints(self.input.len(), checksum, crc32(&self.output), "CHECKSUM");
        }
        
        &self.output
    }
    
    fn update_write(&mut self, i: &mut usize) {
        if self.family == LPRGB_FAMILY_ID && self.block > 0 {
            self.output.append(&mut vec! [0; BLOCK_SIZE_BYTES]);
        }
        
        for j in 0..BLOCK_SIZE_BITS {
            let shift = 6 - j % 7;
            let target = self.block * BLOCK_SIZE_BYTES + j / 8;
            
            if self.family == LPX_FAMILY_ID && target >= self.output.len() {
                self.expected_types = vec![UPDATE_FINISH];
                break;
            }
            
            if j % 8 == 0 {
                self.output[target] = 0;
            }
            
            self.output[target] |= (self.input[*i + j / 7] & (1 << shift)) >> shift << (7 - j % 8);
        }
        
        self.block += 1;
        *i += BLOCK_SIZE_7BITS;
    }
    
    fn verify_byte(&self, i: usize, value: u8, error: &str) -> usize {
        if self.input[i] != value {
            panic!("Parse error at position {:#08x}: expected {} {:#02x}", i, error, value);
        } 
        1
    }
    
    fn verify_byte_array(&self, i: usize, values: &Vec<u8>, error: &str) -> usize {
        for j in 0..values.len() {
            if self.input[i + j] != values[j] {
                panic!("Parse error at position {:#010x}: expected {}", i, error);
            }
        }
        
        values.len()
    }
    
    fn verify_byte_in_array(&self, i: usize, values: &Vec<u8>, error: &str) -> usize {
        if !values.contains(&self.input[i]) {
            panic!(print_vector_error(i, error, values));
        }
        1
    }
    
    fn nibbles_to_u32(&self, i: &mut usize, length: usize) -> u32 {
        let mut result: u32 = 0;
        
        for _ in 0..length {
            result <<= 0x04;
            result |= (self.input[*i] & 0x0F) as u32;
            *i += 1;
        }
        result
    }
}

#[derive(StructOpt)]
struct Cli {
    #[structopt(validator(input_validate), help("Path to input .syx file"))]
    input: PathBuf,
    
    #[structopt(help("Optional output file name. If none is present, output name will be the same as the input but with .bin extension."))]
    output: Option<String> 
}

fn main() {
    let args = Cli::from_args();
    
    let mut path = args.input;
    
    let input_syx = std::fs::read(&path).unwrap();
        
    let out_path = match args.output {
        Some(s) => s,
        None => {
            path.set_extension("bin");
            path.file_name().unwrap().to_str().unwrap().to_string()
        }
    };
        
    match std::fs::write(&out_path, SyxToBin::new(input_syx).convert()) {
        Ok(_) =>  println!("Success! Saved to {}", &out_path),
        Err(err) => panic!(err)
    }
}